spec/jobs/application_job_spec.rb

```diff
@@ -2,34 +2,78 @@
+  def self.lock_timeout
+    1
+  end
+
+  def perform(param_a, param_b, should_raise)
+    Rails.logger.info("simulate a running job to better test queue locking: (#{param_a}, #{param_b}, #{should_raise})")
+
+    raise 'oops' if should_raise
+  end
+end
+
 describe ApplicationJob, type: :job do
+  include ActiveJob::TestHelper
+
   around do |example|
-    old_adapter = described_class.queue_adapter
-    described_class.queue_adapter = :resque
+    old_adapter = ActiveJob::Base.queue_adapter
+    ActiveJob::Base.queue_adapter = :resque
     example.run
-    described_class.queue_adapter = old_adapter
+    ActiveJob::Base.queue_adapter = old_adapter
   end
 
   it 'queues start empty' do
-    expect(Resque.info).to include(pending: 0)
+    expect(enqueued_jobs.size).to eq 0
   end
 
   context 'a subclass with message(s) queued' do
-    before do
-      allow(ZipmakerJob).to receive(:perform_later).and_call_original # undo rails_helper block
-      ZipmakerJob.perform_later('1234abc', 1)
-    end
-
     it 'does not add duplicate messages' do
-      expect { ZipmakerJob.perform_later('1234abc', 1) }
-        .not_to change { Resque.info[:pending] }.from(1)
+      RegularParameterJob.perform_later('1234abc', 1, false)
+      expect { RegularParameterJob.perform_later('1234abc', 1, false) }
+        .not_to change(enqueued_jobs, :size).from(1)
     end
 
     it 'but adds novel messages' do
-      expect { ZipmakerJob.perform_later('7890xyz', 1) } # different druid
-        .to change { Resque.info[:pending] }.from(1).to(2)
-      expect { ZipmakerJob.perform_later('1234abc', 2) } # same druid, different version
-        .to change { Resque.info[:pending] }.from(2).to(3)
+      RegularParameterJob.perform_later('1234abc', 1, false)
+      expect { RegularParameterJob.perform_later('7890xyz', 1, false) } # different druid
+        .to change(enqueued_jobs, :size).from(1).to(2)
+      expect { RegularParameterJob.perform_later('1234abc', 2, false) } # same druid, different version
+        .to change(enqueued_jobs, :size).from(2).to(3)
+    end
+
+    it 'cleans up its lock after succeeding so that the same job with the same params can be queued again immediately' do
+      RegularParameterJob.perform_later('4321cba', 1, false)
+
+      perform_enqueued_jobs
+
+      expect { RegularParameterJob.perform_later('4321cba', 1, false) }
+        .to change(enqueued_jobs, :size).from(0).to(1)
+    end
+
+    it 'cleans up its lock after failing so that the same job with the same params can be queued again immediately' do
+      RegularParameterJob.perform_later('0987zyx', 1, true)
+
+      begin
+        perform_enqueued_jobs
+      rescue # rubocop:disable Style/RescueStandardError terrible pattern in almost all cases, but acceptable here
+        # the error is bubbling up to here despite coming from an async job.  IRL, this would not happen.
+        # no idea why ActiveJob is behaving differently in the tests.
```

Comment:
also, now that i think about this, maybe this comment is off-base.  app code that uses ActiveJob really only ever queues jobs, and doesn't pick them up for execution, and whatever adapter actually runs the job probably does have to deal with this sort of error.  will fix...